wire-compiler/src/main/java/com/squareup/wire/schema/NewSchemaLoader.kt

```diff
@@ -16,118 +16,89 @@
 package com.squareup.wire.schema
 
 import com.google.common.io.Closer
-import com.squareup.wire.schema.internal.parser.ProtoParser
-import okio.buffer
-import okio.source
+import com.squareup.wire.schema.internal.parser.ProtoFileElement
 import java.io.Closeable
 import java.io.IOException
 import java.nio.file.FileSystem
-import java.util.ArrayDeque
+import java.util.TreeSet
 
 /**
  * Load proto files and their transitive dependencies and parse them. Keep track of which files were
  * loaded from where so that we can use that information later when deciding what to generate.
  */
 class NewSchemaLoader(
-  private val fs: FileSystem,
-
-  /** See [com.squareup.wire.WireRun.sourcePath]. */
-  private val sourcePath: List<Location>,
-
-  /** See [com.squareup.wire.WireRun.protoPath]. */
-  private val protoPath: List<Location> = listOf()
-) : Closeable {
+  private val fs: FileSystem
+) : Closeable, Loader {
   private val closer = Closer.create()
 
   /** Errors accumulated by this load. */
   private val errors = mutableListOf<String>()
 
-  /** Files loaded by their relative paths. This is the same path that we use in imports! */
-  private val loaded = mutableMapOf<String, ProtoFile>()
+  /** Paths accumulated that we failed to load. */
+  private val missingImports = TreeSet<String>()
 
-  /** Source files that were loaded. Used to differentiate sources from protoPath elements. */
-  lateinit var sourceLocationPaths: Set<String>
+  /** Source path roots that need to be closed */
+  private lateinit var sourcePathRoots: List<Root>
 
-  /** Working backlog of imported .proto files to load. */
-  private val imports = ArrayDeque<String>()
+  /** Proto path roots that need to be closed */
+  private lateinit var protoPathRoots: List<Root>
 
-  @Throws(IOException::class)
-  fun load(): List<ProtoFile> {
-    check(loaded.isEmpty()) { "do not reuse instances of this class" }
+  /** Initialize the [WireRun.sourcePath] and [WireRun.protoPath] from which files are loaded. */
+  fun initRoots(
+    sourcePath: List<Location>,
+    protoPath: List<Location> = listOf()
+  ) {
+    sourcePathRoots = allRoots(closer, sourcePath)
+    protoPathRoots = allRoots(closer, protoPath)
+  }
 
-    loaded[DESCRIPTOR_PROTO] = loadDescriptorProto()
+  /** Returns the files in the source path. */
+  @Throws(IOException::class)
+  fun loadSourcePathFiles(): List<ProtoFile> {
+    check(::sourcePathRoots.isInitialized && ::protoPathRoots.isInitialized) {
```

Comment:
I recommend going with nullables, it's easy to forget placing an `isInitialized` check.

Reply:
SGTM! Fixed.